delira/training/tf_trainer.py

```diff
@@ -308,3 +314,304 @@ def load_state(self, file_name, *args, **kwargs):
+                         optimizer_params=optimizer_params,
+                         train_metrics=train_metrics,
+                         val_metrics=val_metrics,
+                         lr_scheduler_cls=lr_scheduler_cls,
+                         lr_scheduler_params=lr_scheduler_params,
+                         gpu_ids=gpu_ids,
+                         save_freq=save_freq,
+                         optim_fn=optim_fn,
+                         key_mapping=key_mapping,
+                         logging_type=logging_type,
+                         logging_kwargs=logging_kwargs,
+                         fold=fold,
+                         callbacks=callbacks,
+                         start_epoch=start_epoch,
+                         metric_keys=metric_keys,
+                         convert_batch_to_npy_fn=convert_batch_to_npy_fn,
+                         val_freq=val_freq,
+                         **kwargs
+                         )
+
+        self._setup(network, optim_fn, optimizer_cls, optimizer_params,
+                    lr_scheduler_cls, lr_scheduler_params,
+                    key_mapping, convert_batch_to_npy_fn, gpu_ids)
+
+        for key, val in kwargs.items():
+            setattr(self, key, val)
+
+    def _setup(self, network, optim_fn, optimizer_cls, optimizer_params,
+               lr_scheduler_cls, lr_scheduler_params, key_mapping,
+               convert_batch_to_npy_fn, gpu_ids):
+        """
+        Defines the Trainers Setup
+
+        Parameters
+        ----------
+        network : instance of :class: `AbstractTfNetwork`
+            the network to train
+        optim_fn : function
+            creates a dictionary containing all necessary optimizers
+        optimizer_cls : subclass of tf.train.Optimizer
+            optimizer class implementing the optimization algorithm of choice
+        optimizer_params : dict
+        lr_scheduler_cls : Any
+            learning rate schedule class: must implement step() method
+        lr_scheduler_params : dict
+            keyword arguments passed to lr scheduler during construction
+        convert_batch_to_npy_fn : type, optional
+            function converting a batch-tensor to numpy, per default this is
+            the identity function
+        gpu_ids : list
+            list containing ids of GPUs to use; if empty: use cpu instead
+        """
+
+        if gpu_ids and tf.test.is_gpu_available():
+            self.use_gpu = True
+            if len(gpu_ids) > 1:
+                logger.warning(
+                    "multi-GPU training not yet tested!")
+
+                network = tf.keras.utils.multi_gpu_model(
+                    network,
+                    gpu_ids,
+                    cpu_merge=True,
+                    cpu_relocation=False)
+
+                self.input_device = "/cpu:0"
+                self.output_device = "/cpu:0"
+            else:
+                self.input_device = "/gpu:0"
```

Comment:
Does this always work? What if we want to train on `/gpu:1`?

Reply:
You're right, I should query the first device in the list here.